### ISSUE TITLE
Dimensional contraction support

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#13153]
Sometimes dimensions get contracted.

For example, when setting the speed of light to one, the spacial and temporal dimensions become the same (seconds can be converted to meters by using the speed of light value). Addition of length and time intervals should therefore be allowed (which is currently not allowed).

How would you implement a contraction API? And a de-contraction API to restore the physical constants?

Sample idea for API:
```python
>>> c = speed_of_light
>>> contract_dims(Eq(E, m*c**2), contract_quantities=[c])
E == m
>>> decontract_dims(Eq(E, m), contract_quantities=[c], dim_map={E: energy, m: mass})
E == m*c**2
```

**Comment by Upabjojr** [comment]
@schymans you might be interested.

**Comment by schymans** [comment]
Interesting idea, but I don't understand the example. Could you refer to e.g. a wikipedia article that explains the underlying reasoning? I assume that `E` is a variable with the units of energy, i.e. `M*L*T**(-2)`, but the contraction seems to convert it to units of `M`. Does this imply that `L*T**(-2) = 1`?

**Comment by Upabjojr** [comment]
Yes, the speed of light is the constant that is being contracted, therefore the dimension of speed will become dimensionless, so that `length == time`. At this point the energy has the same dimension as mass.

Some descriptions can be found in: https://en.wikipedia.org/wiki/Natural_units

**Comment by Upabjojr** [comment]
Without going into relativity or other strange stuff, the Gaussian units are a simpler example:
https://en.wikipedia.org/wiki/Gaussian_units

The idea behind the Gaussian unit is simply not to define the _Ampere_, and Coulomb's constant becomes one.

Maybe `convert_to` could be updated to support conversions with dimensional contractions.

**Comment by schymans** [comment]
Thanks for the examples. I think that there is more to this than just unit contraction during conversion. As explained in the wikipedia article on Gaussian units, use of the Gaussian unit system also requires reformulating some fundamental physical equations. Shouldn't the conversion be part of `UnitSystem`? I think the contraction only makes sense when using a specific unit system, e. g. in the Planck system of units ([https://en.wikipedia.org/wiki/Planck_units](https://en.wikipedia.org/wiki/Planck_units), so perhaps someone could define, similarly to `sympy/physics/units/systems/mks.py`, a `planck_units.py`, a `gauss_units.py` etc.

**Comment by Upabjojr** [comment]
> the Gaussian unit system also requires reformulating some fundamental physical equations.

I think that the idea behind the contraction of dimensions is that you take a physical law and drop the constant. In the case of gaussian units, the Coulomb law becomes constantless:
```
F = (1/4 pi epsilon0) q1 q2 / r^2  ===> F = q1 q2 / r^2
```
The dimensions of the constant `(1/4 pi epsilon0)` are `force*length**2*charge**-2`, which is set to unit and then, as an equation, used to abolish the electric current dimension (i.e. all units prop to Ampere).

The gaussian units are then followed by a conversion `meter ==> centimeter` and `kg ==> gram`, but that's the easy part, the hard part remains to determine how the contraction should work (even just creating a good/intuitive API to represent the conversion could be challenging).

**Comment by schymans** [comment]
> `F = (1/4 pi epsilon0) q1 q2 / r^2  ===> F = q1 q2 / r^2`
This implies that the quantities `F`, `q1`, `q2` and `r` have to contain information about the unit system they refer to, right? Otherwise, we would have no chance of knowing that `F = q1 q2 / r^2` is only valid in the Gaussian unit system and vice versa. In sympy/physics/units/definitions.py, we define different quantities for e.g. `mass` and `planck_mass`, so in the above case, we would need to define `gauss_force`, `gauss_charge` and `gauss_distance`? I would advocate to either create, in addition to `Quantity`, a `PhysVariable`, or add to `Quantity` information about the default units. For example, one could then easily define `F`, `q1`, `q2` and `r` as quantities or physical variables with Gauss units as a default and then whatever value the user assigns to them, that value would be converted to Gauss units so that injection into e.g. `F = q1 q2 / r^2` yields the correct result. Would this help?

**Comment by Upabjojr** [comment]
I was considering the basic unit in the gaussian system, the statCoulomb. As of the current state of the code in SymPy, we cannot define it:

```python
statcoulomb = Quantity("statcoulomb", charge, centimeter**(S(3)/2)*gram**(S(1)/2)/second)
```

because we get `ValueError: quantity value and dimension mismatch`. Yet, the _statcoulomb_ measures charges and that's how it is defined.

The dimensional dependencies are registered in a static dictionary inside the class `Dimension`. When I organized it like that I was thinking about the MKS/SI units and followed that kind of dependency. There is a class called _DimensionSystem_, but it is basically not used. Maybe the class _DimensionSystem_ should keep track of the dimensional dependencies and quantity should accept a dimension system parameter:

```python
coulomb = Quantity(
    "coulomb",
    charge,
    ampere*second,
    dimension_system=SI_dimension_system
)

statcoulomb = Quantity(
    "statcoulomb",
    charge,
    centimeter**(S(3)/2)*gram**(S(1)/2)/second,
    dimension_system=gaussian_dimension_system
)
```

At this point the dimensional dependencies will be different between the two dimension systems.

**Comment by schymans** [comment]
This would be awesome. Plus, would it be possible to make `scale_factor` optional in a quantity? Then one could use `Quantity()` to define physical variables with information on the dimension and dimensional dependencies, but not necessarily prescribing a scale factor that affects assumptions attached to that quantity. See also [PR#13268](https://github.com/sympy/sympy/pull/13268).
